src/Workspaces/Core/Portable/SymbolId/SymbolKey.NamedTypeSymbolKey.cs

```diff
@@ -1,77 +1,74 @@
+                visitor.WriteSymbolKey(symbol.ContainingSymbol);
+                visitor.WriteInteger(symbol.Arity);
+                visitor.WriteInteger((int)symbol.TypeKind);
+                visitor.WriteBoolean(symbol.IsUnboundGenericType);
 
-                if (!symbol.Equals(symbol.ConstructedFrom) && !_isUnboundGenericType)
+                if (!symbol.Equals(symbol.ConstructedFrom) && !symbol.IsUnboundGenericType)
+                {
+                    visitor.WriteSymbolKeyArray(symbol.TypeArguments);
+                }
+                else
                 {
-                    _typeArgumentKeysOpt = symbol.TypeArguments.Select(a => GetOrCreate(a, visitor)).ToArray();
+                    visitor.WriteSymbolKeyArray(default(ImmutableArray<ITypeSymbol>));
                 }
             }
 
-            public override SymbolKeyResolution Resolve(Compilation compilation, bool ignoreAssemblyKey, CancellationToken cancellationToken)
+            public static int GetHashCode(GetHashCodeReader reader)
             {
-                var containerInfo = _containerKey.Resolve(compilation, ignoreAssemblyKey, cancellationToken);
-                var types = GetAllSymbols<INamespaceOrTypeSymbol>(containerInfo).SelectMany(s => Resolve(compilation, s, ignoreAssemblyKey));
-                return CreateSymbolInfo(types);
+                return Hash.Combine(reader.ReadString(),
+                       Hash.Combine(reader.ReadSymbolKey(),
+                       Hash.Combine(reader.ReadInteger(),
+                       Hash.Combine(reader.ReadInteger(),
+                       Hash.Combine(reader.ReadBoolean(),
+                                    reader.ReadSymbolKeyArrayHashCode())))));
             }
 
-            private IEnumerable<INamedTypeSymbol> Resolve(
-                Compilation compilation,
-                INamespaceOrTypeSymbol container,
-                bool ignoreAssemblyKey)
+            public static SymbolKeyResolution Resolve(SymbolKeyReader reader)
             {
-                var types = container.GetTypeMembers(GetName(_metadataName), _arity);
-                var result = InstantiateTypes(compilation, ignoreAssemblyKey, types, _arity, _typeArgumentKeysOpt);
+                var metadataName = reader.ReadString();
+                var containingSymbolResolution = reader.ReadSymbolKey();
+                var arity = reader.ReadInteger();
+                var typeKind = (TypeKind)reader.ReadInteger();
+                var isUnboundGenericType = reader.ReadBoolean();
+                var typeArgumentsOpt = reader.ReadSymbolKeyArray();
 
-                return _isUnboundGenericType
-                    ? result.Select(t => t.ConstructUnboundGenericType())
-                    : result;
+                var types = GetAllSymbols<INamespaceOrTypeSymbol>(containingSymbolResolution).SelectMany(
+                    s => Resolve(reader, s, metadataName, arity, typeKind, isUnboundGenericType, typeArgumentsOpt));
```

Comment:
This statement could be unrolled to get rid of extraneous allocations and delegate captures.